protocols/identify/src/handler.rs

```diff
@@ -90,36 +128,81 @@ pub struct Handler {
 
     /// The interval of `trigger_next_identify`, i.e. the recurrent delay.
     interval: Duration,
+
+    /// The public key of the local peer.
+    public_key: PublicKey,
+
+    /// Application-specific version of the protocol family used by the peer,
+    /// e.g. `ipfs/1.0.0` or `polkadot/1.0.0`.
+    protocol_version: String,
+
+    /// Name and version of the peer, similar to the `User-Agent` header in
+    /// the HTTP protocol.
+    agent_version: String,
+
+    /// Address observed by or for the remote.
+    observed_addr: Multiaddr,
 }
 
-/// Event produced by the `IdentifyHandler`.
+/// Information provided by the `Behaviour` upon requesting.
+#[derive(Debug)]
+pub struct BehaviourInfo {
+    /// The addresses that the peer is listening on.
+    pub listen_addrs: Vec<Multiaddr>,
+
+    /// The list of protocols supported by the peer, e.g. `/ipfs/ping/1.0.0`.
+    pub protocols: Vec<String>,
+}
+
+/// Event produced by the `Handler`.
 #[derive(Debug)]
 #[allow(clippy::large_enum_variant)]
 pub enum Event {
     /// We obtained identification information from the remote.
     Identified(Info),
+    /// We replied to an identification request from the remote.
+    Identification(PeerId),
     /// We actively pushed our identification information to the remote.
     IdentificationPushed,
     /// We received a request for identification.
-    Identify(ReplySubstream<NegotiatedSubstream>),
-    /// Failed to identify the remote.
+    Identify,
+    /// Failed to identify the remote, or to reply to an identification request.
     IdentificationError(ConnectionHandlerUpgrErr<UpgradeError>),
 }
 
-/// Identifying information of the local node that is pushed to a remote.
 #[derive(Debug)]
-pub struct Push(pub Info);
+#[allow(clippy::large_enum_variant)]
+pub enum InEvent {
+    /// Identifying information of the local node that is pushed to a remote.
+    Push(Info),
+    /// Identifying information requested from this node.
+    Identify(BehaviourInfo),
```

Comment:
How different are these two really? Isn't exactly the same, just with different semantics?

Reply:
I wasn't talking about the size difference actually.

Semantically, even for an identify push, all the handler needs to know are the updated protocols and listen addresses, which is represented as the `BehaviourInfo` struct. Can we model it as the same type?

Reply:
ah right! So I followed along and merge `Behaviour`'s `pending_push` into `requests` for better balance when fullfiling them. Previously `Push` requests would always have priority over `Identify` ptal Thomas.